src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,188 @@
+        Node<K, V> existingNode = getNode(key);
+        if (existingNode != null) {
+            existingNode.value = value;
+        } else {
+            Node<K, V> newNode = new Node<>(hashOfKey, key, value, null);
+            insertNode(newNode, bucketIndex);
+            size++;
+        }
     }
 
     @Override
     public V getValue(K key) {
+        Node<K, V> node = getNode(key);
+        return node == null ? null : node.value;
+    }
+
+    private Node<K, V> getNode(K key) {
+        int hashOfKey = calculateHash(key);
+        int bucketIndex = getBucketIndex(hashOfKey);
+        if (table[bucketIndex] == null) {
+            return null;
+        }
+        Node<K, V> current = table[bucketIndex];
+        while (current != null) {
+            if (current.hash == hashOfKey
+                    && (current.key == key || Objects.equals(current.key, key))) {
+                return current;
+            }
+            current = current.next;
+        }
         return null;
     }
 
+    private void insertNode(Node<K, V> newNode, int bucketIndex) {
+        if (table[bucketIndex] == null) {
+            table[bucketIndex] = newNode;
+        } else {
+            getLastNode(bucketIndex).next = newNode;
+        }
+    }
+
+    private Node<K, V> getLastNode(int bucketIndex) {
+        Node<K, V> current = table[bucketIndex];
+        while (current.next != null) {
+            current = current.next;
+        }
+        return current;
+    }
+
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    @Override
+    public V remove(K key) {
+        Node<K, V> neededNode = getNode(key);
+        if (neededNode == null) {
+            return null;
+        }
+        int bucketIndex = getBucketIndex(neededNode.hash);
+        Node<K, V> previous = null;
+        Node<K, V> current = table[bucketIndex];
+        while (current != null) {
+            if (Objects.equals(neededNode, current)) {
+                V oldValue = current.value;
+                if (previous != null) {
+                    previous.next = current.next;
+                } else {
+                    table[bucketIndex] = current.next;
+                }
+                size--;
+                return oldValue;
+            }
+            previous = current;
+            current = current.next;
+        }
+        return null;
+    }
+
+    @Override
+    public boolean containsKey(K key) {
+        return getNode(key) != null;
+    }
+
+    @Override
+    public boolean containsValue(V value) {
+        for (Node<K, V> node : table) {
+            Node<K, V> current = node;
+            while (current != null) {
+                if (Objects.equals(current.value, value)) {
+                    return true;
+                }
+                current = current.next;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return getSize() == 0;
```

Comment:
```suggestion
        return size == 0;
```

Reply:
Should I remove the hash variable only from the constructor or from the Node class too?
If only from the constructor, the assignment to the hash variable should look like this?
Node<K, V> newNode = new Node<>(key, value, null);
newNode.hash = hashOfKey;